src/Http/Http.Abstractions/src/PathString.cs

```diff
@@ -187,9 +190,12 @@ public static PathString FromUriComponent(Uri uri)
             {
                 throw new ArgumentNullException(nameof(uri));
             }
-
-            // REVIEW: what is the exactly correct thing to do?
-            return new PathString("/" + uri.GetComponents(UriComponents.Path, UriFormat.Unescaped));
+            var uriComponent = uri.GetComponents(UriComponents.Path, UriFormat.UriEscaped);
+            Span<char> pathBuffer = new char[uriComponent.Length + 1];
```

Comment:
Feels like this could be a stackalloc for 90% of the cases:

```suggestion
            const int stackAllocThreshold = 128;
            Span<char> pathBuffer = uriComponent.Length < stackAllocThreshold ? stackalloc char[stackAllocThreshold] : new char[uriComponent.Length + 1];
```

Reply:
I was just thinking about it, and I was worried on not having a good maximum. I will do this with stackallow and do some perf tests.

Reply:
Performance Results:
```
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19043.1052 (21H1/May2021Update)
Intel Core i5-1035G4 CPU 1.10GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.100-preview.5.21264.3
  [Host]     : .NET 6.0.0 (6.0.21.30701), X64 RyuJIT
  Job-YAOZWJ : .NET 6.0.0 (6.0.21.26301), X64 RyuJIT

Server=True  Toolchain=.NET Core 6.0  RunStrategy=Throughput

|                 Method |            testPath |              testUri |      Mean |     Error |    StdDev |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |-------------------- |--------------------- |----------:|----------:|----------:|-------------:|-------:|------:|------:|----------:|
| OnPathFromUriComponent |        /api/a%2Fb/c |                    ? |  65.18 ns |  1.288 ns |  2.451 ns | 15,342,606.8 | 0.0018 |     - |     - |      48 B |
| OnPathFromUriComponent | /thi(...)%20b [156] |                    ? | 349.51 ns | 10.863 ns | 31.516 ns |  2,861,188.7 | 0.0248 |     - |     - |     672 B |
|  OnUriFromUriComponent |                   ? | https(...)2Fb/c [35] | 142.56 ns |  2.860 ns |  4.857 ns |  7,014,802.6 | 0.0033 |     - |     - |      96 B |
|  OnUriFromUriComponent |                   ? |  http(...)/a b [177] | 736.39 ns | 24.223 ns | 71.423 ns |  1,357,978.5 | 0.0372 |     - |     - |   1,016 B |
```

Reply:
I'm not sure how to compare these to the numbers above.

Reply:
The 1st and 3rd rows are the ones that may be compared (because they work on the same inputs). The 2nd and 4th rows measure the performance of an input which is over the stackalloc limit (they have no baseline measurement so far).  I'll prepare a before-after perf tests later today.

Reply:
```
// * Summary *

BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19043.1052 (21H1/May2021Update)
Intel Core i5-1035G4 CPU 1.10GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.100-preview.5.21264.3
  [Host]     : .NET 6.0.0 (6.0.21.30701), X64 RyuJIT
  Job-URULBA : .NET 6.0.0 (6.0.21.26301), X64 RyuJIT

Server=True  Toolchain=.NET Core 6.0  RunStrategy=Throughput

|                        Method |            testPath |              testUri |      Mean |     Error |    StdDev |    Median |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------ |-------------------- |--------------------- |----------:|----------:|----------:|----------:|-------------:|-------:|------:|------:|----------:|
|  OnPathFromUriComponent_After |        /api/a%2Fb/c |                    ? |  62.27 ns |  1.192 ns |  2.488 ns |  62.38 ns | 16,057,996.3 | 0.0017 |     - |     - |      48 B |
| OnPathFromUriComponent_Before |        /api/a%2Fb/c |                    ? |  49.62 ns |  1.062 ns |  2.872 ns |  49.26 ns | 20,154,226.7 | 0.0017 |     - |     - |      48 B |
|  OnPathFromUriComponent_After | /thi(...)%20b [156] |                    ? | 380.84 ns | 15.064 ns | 44.418 ns | 376.84 ns |  2,625,785.1 | 0.0253 |     - |     - |     672 B |
| OnPathFromUriComponent_Before | /thi(...)%20b [156] |                    ? |  78.30 ns |  1.779 ns |  5.132 ns |  76.66 ns | 12,771,165.7 | 0.0125 |     - |     - |     336 B |
|   OnUriFromUriComponent_After |                   ? | https(...)2Fb/c [35] | 150.79 ns |  3.081 ns |  5.397 ns | 150.91 ns |  6,631,899.3 | 0.0036 |     - |     - |      96 B |
|  OnUriFromUriComponent_Before |                   ? | https(...)2Fb/c [35] | 134.99 ns |  2.719 ns |  4.833 ns | 134.66 ns |  7,407,887.2 | 0.0033 |     - |     - |      96 B |
|   OnUriFromUriComponent_After |                   ? |  http(...)/a b [177] | 752.86 ns | 22.151 ns | 64.265 ns | 742.55 ns |  1,328,273.2 | 0.0381 |     - |     - |   1,016 B |
|  OnUriFromUriComponent_Before |                   ? |  http(...)/a b [177] | 719.44 ns | 14.336 ns | 40.200 ns | 716.70 ns |  1,389,965.7 | 0.0248 |     - |     - |     672 B |
```

Reply:
@davidfowl I found the reason why the longer string input is significantly slower (380.84 ns vs 78.30 ns). 
It seems the previous implementation using [Uri.UnescapeDataString](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Private.Uri/src/System/UriExt.cs#L542) starts decoding from the first occurrence of % char
In my benchmark, I had the first % char only at the very end of the benchmarked input string: `/thisMustBeAVeryLongPath/SoLongThatItCouldActuallyBeLargerToTheStackAllocThresholdValue/PathsShorterToThisAllocateLessOnHeapByUsingStackAllocation/api/a%20b`

Having a % early in the input string drops *OnPathFromUriComponent_Before* performance for the longer string from ~78ns to ~300ns on the my machine.

I am wondering about the following: the new `Span<char>` overload in UrlDecoder supports an isFormEncoding, but it is not used. Should I keep the signature of the `Span<byte>` and `Span<char>` the same, or may I drop this unused feature for the new `Span<char>` overload? This would leave space to add the same optimization to the `Span<char>` overload on UrlDecoder as pointed out in [Uri.UnescapeDataString](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Private.Uri/src/System/UriExt.cs#L542).
Or shall I keep isFormEncoding and work around it?

Reply:
> Should I keep the signature of the `Span<byte>` and `Span<char>` the same, or may I drop this unused feature for the new `Span<char>` overload?

We had initially wanted them to stay the same for maintenance purposes, but if you have a compelling reason to diverge then do it. Leave a comment on the new API that form decoding is not supported and mention the differences.

Reply:
I have updated the review, and re-run performance tests. I have also extended the test with a new case, a longer path but still under stackalloc limit, which contains a % encoded value in the beginning.

Test cases:
- _Before _After denotes before or after the change.
- 3 `string` testPaths tested on the *string* overload of PathString.FromUriComponent, a shorter one and 2 longer one. These input strings can be found in `PathStringBenchmark` in this PR.
- 3 `Uri` testUri tested on the *Uri* overload of PathString.FromUriComponent. The Uri are the testPaths prefixed with scheme, base address and port.

```
// * Summary *

BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19043.1052 (21H1/May2021Update)
Intel Core i5-1035G4 CPU 1.10GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.100-preview.5.21264.3
  [Host]     : .NET 6.0.0 (6.0.21.30701), X64 RyuJIT
  Job-KIGUYA : .NET 6.0.0 (6.0.21.26301), X64 RyuJIT

Server=True  Toolchain=.NET Core 6.0  RunStrategy=Throughput

|                        Method |            testPath |              testUri |      Mean |     Error |    StdDev |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------ |-------------------- |--------------------- |----------:|----------:|----------:|-------------:|-------:|------:|------:|----------:|
|  OnPathFromUriComponent_After |        /api/a%2Fb/c |                    ? |  52.49 ns |  1.017 ns |  1.044 ns | 19,050,746.4 | 0.0018 |     - |     - |      48 B |
| OnPathFromUriComponent_Before |        /api/a%2Fb/c |                    ? |  45.38 ns |  0.962 ns |  2.342 ns | 22,036,498.6 | 0.0017 |     - |     - |      48 B |
|  OnPathFromUriComponent_After | /t%2(...)%20b [125] |                    ? | 241.71 ns |  4.816 ns |  9.393 ns |  4,137,201.9 | 0.0098 |     - |     - |     264 B |
| OnPathFromUriComponent_Before | /t%2(...)%20b [125] |                    ? | 287.73 ns |  5.793 ns | 15.160 ns |  3,475,448.6 | 0.0095 |     - |     - |     264 B |
|  OnPathFromUriComponent_After | /thi(...)%20b [156] |                    ? |  85.37 ns |  1.466 ns |  1.224 ns | 11,714,264.0 | 0.0249 |     - |     - |     672 B |
| OnPathFromUriComponent_Before | /thi(...)%20b [156] |                    ? |  79.77 ns |  2.813 ns |  8.024 ns | 12,536,382.9 | 0.0125 |     - |     - |     336 B |
|   OnUriFromUriComponent_After |                   ? | https(...)2Fb/c [35] | 150.69 ns |  2.221 ns |  2.078 ns |  6,636,159.4 | 0.0033 |     - |     - |      96 B |
|  OnUriFromUriComponent_Before |                   ? | https(...)2Fb/c [35] | 151.91 ns |  2.206 ns |  2.064 ns |  6,582,884.8 | 0.0036 |     - |     - |      96 B |
|   OnUriFromUriComponent_After |                   ? |  http(...)/a b [144] | 580.26 ns | 11.667 ns | 24.864 ns |  1,723,372.0 | 0.0200 |     - |     - |     544 B |
|  OnUriFromUriComponent_Before |                   ? |  http(...)/a b [144] | 638.59 ns | 12.364 ns | 19.966 ns |  1,565,939.6 | 0.0191 |     - |     - |     536 B |
|   OnUriFromUriComponent_After |                   ? |  http(...)/a b [177] | 517.59 ns | 12.086 ns | 34.286 ns |  1,932,038.2 | 0.0372 |     - |     - |   1,016 B |
|  OnUriFromUriComponent_Before |                   ? |  http(...)/a b [177] | 760.50 ns | 14.819 ns | 20.774 ns |  1,314,918.3 | 0.0248 |     - |     - |     672 B |
```

I am more comfortable with these numbers.
I notice (maybe a bit late) but there are some bit shifting in UrlDecoder. After a quick search I see Kestrel not supporting bigendian and throws PlatformNotSupportedException, so I assume that should be OK.